Add mariaDB4j-springboot module for auto-configuration with spring boot

kbyyd24 opened the pull request:
Add a module named mariaDB4j-springboot, it will auto-configure `MariaDBSpringService` as a bean named `mariaDB4j`, and initialize bean `dataSource` after `mariaDB4j` initialized by annotation `@DependsOn`.
So now, we can use mariaDB4j with spring boot with dependency of mariaDB4j-springboot. For example, we can configure gradle like this:

```gradle
dependencies {
    compile 'org.springframework.boot:spring-boot-starter'
    testCompile 'ch.vorburger.mariaDB4j:mariaDB4j-springboot:2.3.1-SNAPSHOT'
}
```

duttonw commented:
can you update the readme in the root on how to use this as well as create an example integration project to has sample config setup with spring boot. (how you will use this in a simple spring-boot app)
see https://github.com/spring-projects/spring-boot/tree/master/spring-boot-samples/spring-boot-sample-flyway but for MariaDB4j instead.

kbyyd24 commented:
@duttonw I will update the readme in the root. But it's necessary to create another example integration project to has sample config setup with spring boot? I think mariaDB4j-app is a sample of it, maybe what I need to do is update it. Is it OK to change mariaDB4j-app?

vorburger commented:
>  I think mariaDB4j-app is a sample of it, maybe what I need to do is update it. Is it OK to change mariaDB4j-app?

That makes sense to me, yes - go for it.

vorburger commented:
@kbyyd24 if you work @duttonw feedback into this and resolve the minor new merge conflict, I'd review this.

kbyyd24 commented:
@vorburger I updated readme, almost like @duttonw feedback, the only thing different is the link to the  example application, I changed it to [mariaDB4j-app](https://github.com/vorburger/MariaDB4j/blob/master/mariaDB4j-app/). I Also resolved the conflict. So I think it's ready to be review.

kbyyd24 commented:
@duttonw Thanks for your feedback about the readme and my blog. I have updated the readme as you feedback. For the JHipster, I had never heard about it, I will try it for a demo. Thanks for your recommend.
By the way, how can you read my blog? It Chinese, do you know that?

duttonw commented:
Hi @kbyyd24 ,

Google chrome allows language translation with a click of a button or [like this](https://translate.google.com/translate?sl=auto&tl=en&js=y&prev=_t&hl=en&ie=UTF-8&u=http%3A%2F%2Fblog.gaoyuexiang.cn%2F2017%2F10%2F25%2FCustome_Validation_in_Java%2F&edit-text=)

kbyyd24 commented:
@duttonw Oh, I forgot it :joy:

kbyyd24 commented:
Hi @duttonw , I don't know how to do that. I clicked the "Update branch" button, then triggered travis automatically. After pipeline passed, this page looked like the picture. Is the rebase done?

![image](https://user-images.githubusercontent.com/9191275/44438472-cc69fe80-a5f1-11e8-862b-9c557c022542.png)

duttonw commented:
Cool, now it just needs @vorburger to do a final review and merge :D :+1:

kbyyd24 commented:
That's cool, Thank you for your help 👍

vorburger commented on mariaDB4j-app/pom.xml:
```diff
@@ -43,6 +43,22 @@
+		<dependency>
+			<groupId>javax.xml.bind</groupId>
+			<artifactId>jaxb-api</artifactId>
+			<version>2.3.0</version>
```
why is this needed here?

vorburger commented on mariaDB4j-app/pom.xml:
```diff
@@ -43,6 +43,22 @@
+		<dependency>
+			<groupId>org.mariadb.jdbc</groupId>
+			<artifactId>mariadb-java-client</artifactId>
+			<version>2.2.6</version>
```
I doubt this is needed, have you tried without it?

Note it currently works without this.... 

The parent/root POM already has this dependency.

Why fix a more recent version? If it were absolutely needed (I doubt it, easier to inherit it from the Spring BOM instead of having to keep bumping it here @duttonw), why only here, not in the parent?

kbyyd24 replied:
I can't find the dependency in the parent/root POM. I can only find it in `mariaDB4j` module and it's scope is test. So need I add the dependency in the parent/root POM?

vorburger replied:
see #164

vorburger commented on mariaDB4j-app/src/main/java/ch/vorburger/mariadb4j/springframework/boot/MariaDB4jApplication.java:
```diff
@@ -63,6 +62,7 @@ public static void main(String[] args) throws Exception {
 
     @Override
     public int getExitCode() {
+        MariaDB4jSpringService dbService = applicationContext.getBean(MariaDB4jSpringService.class);
```
it's been many years since I last used the Spring Framework myself (doing other things nowadays), but wasn't the idea that explicitly looking up beans is bad? I thought it was much better DI practice to just inject what you needed. So why change the ` @Autowired MariaDB4jSpringService` to an `applicationContext.getBean(MariaDB4jSpringService.class)` ?

vorburger commented on mariaDB4j-app/src/main/resources/application.yml:
```diff
@@ -0,0 +1,9 @@
+    password:
+
+mariaDB4j:
+  port: 3307
```
hard-coding to port 3307 means this fails if one runs this on a system where there already is a mysqld running (e.g. from a RPM / DEB launched by systemd).  The original `mariaDB4j.port=0` thing let it automatically choose a free port - that was cool. Is it possible to retain a similar solution? Perhaps it's even enough to simply try out `port: 0` here?

duttonw replied:
Hi @kbyyd24 , have you looked at how in-memory db's get their dynamic port?

kbyyd24 replied:
@duttonw No, I don't know how spring get that. I will research about that.

kbyyd24 replied:
Hi @duttonw , I found that h2, hsql and derby are auto configured in Spring Boot, and they are in-memory DB. So they are running in the same process with Spring Boot. Then we don't need any port to connect to them.
In this case, mariaDB4j is not in-memory DB, so we need to determine a port. I'm going to change the logic of creation dataSource bean. Maybe we don't need use property of `spring.data` here, we can use `mariaDB4j.port`, `mariaDB4j.db-host`, `mariaDB4j.username` and etc.

jordanms replied:
You should use a random port and then set the chosen port in the [datasource](https://github.com/vorburger/MariaDB4j/pull/153/files#diff-5bc8877ceebd71b5302b7e32eb8c130613b238b1f6a75a0aa84161ca267b3a7cR41).

vorburger commented on mariaDB4j-springboot/src/main/java/ch/vorburger/mariadb4j/springboot/autoconfigure/DataSourceAutoConfiguration.java:
```diff
@@ -0,0 +1,46 @@
+ * #%L
+ * MariaDB4j
+ * %%
+ * Copyright (C) 2012 - 2014 Michael Vorburger
```
nit picking, sorry but could you put "2012 - 2018" here?

kbyyd24 replied:
Thank you for your mention, I had never notice this 😂.
Now it fixed.

vorburger commented on mariaDB4j-springboot/src/main/java/ch/vorburger/mariadb4j/springboot/autoconfigure/MariaDB4jSpringConfiguration.java:
```diff
@@ -0,0 +1,36 @@
+ * #%L
+ * MariaDB4j
+ * %%
+ * Copyright (C) 2012 - 2014 Michael Vorburger
```
2018

kbyyd24 replied:
Fixed.

jordanms commented on mariaDB4j-springboot/src/main/java/ch/vorburger/mariadb4j/springboot/autoconfigure/DataSourceAutoConfiguration.java:
```diff
@@ -0,0 +1,46 @@
+    public DataSource dataSource(DataSourceProperties dataSourceProperties) {
+        return DataSourceBuilder.create()
+                .driverClassName(dataSourceProperties.getDriverClassName())
+                .url(dataSourceProperties.getUrl())
```
You should not use the URL coming from the configuration because you will use the production config if it is not overwritten in the test config. You should set it to `jdbc:mariadb://localhost:[dynamic_chosen_port]/test`.

jordanms replied:
Of course, has also to consider the `mariaDB4j` configuration...

jordanms commented on mariaDB4j-springboot/src/main/java/ch/vorburger/mariadb4j/springboot/autoconfigure/DataSourceAutoConfiguration.java:
```diff
@@ -0,0 +1,46 @@
+        return DataSourceBuilder.create()
+                .driverClassName(dataSourceProperties.getDriverClassName())
+                .url(dataSourceProperties.getUrl())
+                .username(dataSourceProperties.getUsername())
```
Same here. You should usee `root`

jordanms commented on mariaDB4j-springboot/src/main/java/ch/vorburger/mariadb4j/springboot/autoconfigure/DataSourceAutoConfiguration.java:
```diff
@@ -0,0 +1,46 @@
+                .driverClassName(dataSourceProperties.getDriverClassName())
+                .url(dataSourceProperties.getUrl())
+                .username(dataSourceProperties.getUsername())
+                .password(dataSourceProperties.getPassword())
```
Same here. You should use `` (empty).

vorburger commented:
@jordanms You would need to open a new PR with suggested changes, or a new issue describing a problem, or suggesting an enhancement. Comments on a 4 year old closed PR are ignored.

duttonw commented:
Hi @jordanms , 

To your comment that you were providing feedback. You need to be in the moment and that moment has past 4 years ago. 

Yes this is an opinionated way of doing it as spring boot is 'very opinionated', sometimes that opinion changes direction and falls out of flavour, but it still is a way of working. 

If you don't like how it was done in the past. Then contribute, change it in a new Pull Request or Raise an Issue ticket as an enhancement request on how to bring it into line with current day spring boot.  

I won't be contributing at this time as you have not articulated in a new ticket on how it could change in a manner that would compel another to implement it.

Regards,

@duttonw 

p.s. This could now be done with conditionals properties which would only enable this class if ```mariaDB4j.port``` property is not found. https://www.baeldung.com/spring-conditional-annotations#predefined-conditional-annotations

jordanms commented:
Hi, @duttonw 

> If you don't like how it was done in the past.

It's not that I don't like it, it just doesn't work. I spent hours of my day trying to figure out how to make it work. The worse thing is that before I try to use the library, I checked if it was maintained and the last update was a few days ago. So I decided to try it. 

When I say that it doesn't work, is because the idea of the feature is that you have autoconfiguration in the way of H2 for testing, and this is not done and has never been achieved from what I saw in the discussions.

I only commented because I thought it could be helpful. I'm not using the library because it uses a very old MariaDB (10.2 I think) which is not supported by Flyway.

Regards
